Emby.Server.Implementations/Library/Validators/CollectionValidator.cs

```diff
@@ -0,0 +1,166 @@
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Collections.Generic;
+using MediaBrowser.Controller.Collections;
+using MediaBrowser.Controller.Entities;
+using MediaBrowser.Controller.Entities.Movies;
+using MediaBrowser.Controller.Library;
+using MediaBrowser.Model.Querying;
+using Microsoft.Extensions.Logging;
+using Jellyfin.Data.Enums;
+
+namespace Emby.Server.Implementations.Library.Validators
+{
+    /// <summary>
+    /// Class CollectionValidator.
+    /// </summary>
+    public class CollectionValidator
+    {
+        /// <summary>
+        /// The library manager.
+        /// </summary>
+        private readonly ILibraryManager _libraryManager;
+
+        /// <summary>
+        /// The collection manager.
+        /// </summary>
+        private readonly ICollectionManager _collectionManager;
+
+        /// <summary>
+        /// The logger.
+        /// </summary>
+        private readonly ILogger<CollectionValidator> _logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CollectionValidator" /> class.
+        /// </summary>
+        /// <param name="libraryManager">The library manager.</param>
+        /// <param name="collectionManager">The collection manager.</param>
+        /// <param name="logger">The logger.</param>
+        public CollectionValidator(ILibraryManager libraryManager, ICollectionManager collectionManager, ILogger<CollectionValidator> logger)
+        {
+            _libraryManager = libraryManager;
+            _collectionManager = collectionManager;
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Runs the specified progress.
+        /// </summary>
+        /// <param name="progress">The progress.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>Task.</returns>
+        public async Task Run(IProgress<double> progress, CancellationToken cancellationToken)
+        {
+            var movies = _libraryManager.GetItemList(new InternalItemsQuery
```

Comment:
I would probably paginate this query so it doesn’t pull all of the movies at once. It doesn’t need to be a small page query, 1,000 is probably good enough

Reply:
May I ask how to paginate this? I need to get all items which type is movie, the InternalItemsQuery don't have skip option. And if I paginate based itemIds, I have to get all movie's itemdId, this need to query all movies at once too. Thanks.

Reply:
There's a `StartIndex` parameter, which is used as an offset when querying.

Reply:
Hi @crobibero , because I need to count how many movies in same collection, I have to store all movies to a map. It makes paginate movie query meanless. And I think the cost is more or less aceeptable, I have nearly 3000 movies, during first time validate , the memory rise about 100MB.

Reply:
Improved a little, now query movies one time per library.